internal/subscriber/subscribe.go

```diff
@@ -74,66 +74,57 @@ func (s *SubscriberService) Start(ctx context.Context) error {
 	return nil
 }
 
-func nackLog(msg *pubsub.Message, err error) {
-	log.Printf("Subscriber error: %v\n", err)
-	msg.Nack() // TODO: Improve error handling by adding proper logging
-}
-
-// consumeMessage is a callback function for pubsub.Receive() which performs
-// updates to database according to incoming metadata.
+// processMessage handles incoming metadata and performs database operations based on the eventType of incoming *pubsub.Message
 //
-// Messages are expected to be unordered so the handling of incoming metadata has to
+// Messages are expected to be unordered. The handling of incoming metadata has to
 // be based on its update time and gracefully Nack()'d when necessary
-func (s *SubscriberService) consumeMessage(ctx context.Context, msg *pubsub.Message) {
+func processMessage(s *SubscriberService, msg *pubsub.Message) error {
+	// parse payload
 	var p payload
 	if err := json.Unmarshal(msg.Data, &p); err != nil {
-		nackLog(msg, err)
-		return
+		return err
 	}
 
 	inMetadata, err := newMetadata(p)
 	if err != nil {
-		nackLog(msg, err)
-		return
+		return err
 	}
 
 	_, isReplaced := msg.Attributes["overwrittenByGeneration"]
 	eventType := msg.Attributes["eventType"]
 
+	// Ignore replacement messages
+	if isReplaced {
+		return nil
+	}
+
 	switch eventType {
 	case storage.ObjectFinalizeEvent:
-		if err := s.handleFinalize(inMetadata); err != nil {
-			nackLog(msg, err)
-			return
+		if err = s.handleFinalize(inMetadata); err != nil {
+			return err
 		}
-
 	case storage.ObjectDeleteEvent:
-		// Ignore replacement events
-		if isReplaced {
-			msg.Ack()
-			return
-		}
-
 		if err := s.handleDelete(inMetadata); err != nil {
-			nackLog(msg, err)
-			return
+			return err
 		}
 	case storage.ObjectArchiveEvent:
-		// Ignore replacement events
-		if isReplaced {
-			msg.Ack()
-			return
-		}
-
 		if err := s.handleArchive(inMetadata); err != nil {
-			nackLog(msg, err)
-			return
+			return err
 		}
 	default:
-		defaultErr := fmt.Errorf("unknown event type: %s", eventType)
-		nackLog(msg, defaultErr)
+		return fmt.Errorf("unknown event type: %s", eventType)
 	}
 
+	return nil
+}
+
+// consumeMessage is a callback function for pubsub.Receive() which handles
+// the acknowledgment of messages based on processMessage() results
+func (s *SubscriberService) consumeMessage(ctx context.Context, msg *pubsub.Message) {
+	if err := processMessage(s, msg); err != nil {
+		log.Printf("message not acknowledged: %v\n", err)
+		msg.Nack()
```

Comment:
no return? can a message be nacked then acked?

Reply:
It is not necessary as calls to `Nack()` or `Ack()` have no effect after the [first call](https://pkg.go.dev/cloud.google.com/go@v0.115.1/internal/pubsub#Message.Nack)

Reply:
I think one of the following two things should happen:
- a). return
- b). add a comment above ack explaining why it's ok to ack after nack

Reply:
Sure! This functionality is not so obvious unless you dig into the docs and I don't think a comment to justify is worth it. Will go for option A!